### ISSUE TITLE
Task property validation file is generated outside of project.buildDir when java-gradle-plugin is applied

### ISSUE THREAD

**Issue by assimd** [gradle/gradle#3546]
When using `java-gradle-plugin`, task property validation file `reports/task-properties/report.txt` is always generated at `build/reports/task-properties/report.txt`, even if `project.buildDir` is updated to point somewhere else.

### Expected Behavior
Task property validation file should be generated at `${project.buildDir}/reports/task-properties/report.txt`

### Current Behavior
Task property validation file should is generated at `build/reports/task-properties/report.txt`

### Context
We use `build/` folder to output the final artifact which are meant to to be archived. When we have `java-gradle-plugin` applied and `project.buildDir` to pointing to somewhere else, we are still seeing `build/reports/task-properties/report.txt` being generated and getting archived along with the other artifacts that we manually copy to `build/` directory.

This seems to be due to [this line](https://github.com/gradle/gradle/blame/a1a32e5877c852b3a99f337e0fc07a2c979840a6/subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java#L250-L252) which sets the `validator.setOutputFile` using the default value of `project.buildDir` instead of setting it lazily.

### Steps to Reproduce (for bugs)
[javaGradlePluginSample.tar.gz](https://github.com/gradle/gradle/files/1489890/javaGradlePluginSample.tar.gz)

Following is the directory structure after the build: 
```
.
├── build
│   └── reports
│       └── task-properties
│           └── report.txt
├── build.gradle
├── fooBar
│   ├── classes
│   │   └── java
│   │       └── main
│   │           └── org
│   │               └── gradle
│   │                   └── sample
│   │                       └── SimplePlugin.class
│   ├── libs
│   │   └── javaGradlePlugin.jar
│   ├── pluginDescriptors
│   │   └── org.gradle.sample.simple-plugin.properties
│   ├── pluginUnderTestMetadata
│   │   └── plugin-under-test-metadata.properties
│   ├── resources
│   │   └── main
│   │       └── META-INF
│   │           └── gradle-plugins
│   │               └── org.gradle.sample.simple-plugin.properties
│   └── tmp
│       ├── compileJava
│       └── jar
│           └── MANIFEST.MF
└── src
    └── main
        └── java
            └── org
                └── gradle
                    └── sample
                        └── SimplePlugin.java

```
Note: There should not be any `build/` directory when `project.buildDir = 'fooBar'`

### Your Environment
 * Build scan URL: https://scans.gradle.com/s/h6ajwgxvlnq26/

**Comment by lptr** [comment]
Thanks for the detailed bug report @assimd, I'll take a look today.

**Comment by blindpirate** [comment]
Thanks for reporting, it's the expected behaviour.

When `java-gradle-plugin` is applied, your code `project.buildDir='fooBar'` hasn't been executed yet. You could use one of the following workarounds:

- Set buildDir before applying the plugin:

```
project.buildDir='fooBar'
apply plugin:'java-gradle-plugin'
```

- Use init.script:

```
allprojects { project ->
    project.buildDir='fooBar'
}
```

**Comment by lptr** [comment]
This will be shipped in Gradle 4.5.
